x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -465,6 +466,12 @@ protected Clock getClock() {
 
         securityIndex.set(SecurityIndexManager.buildSecurityIndexManager(client, clusterService, SECURITY_MAIN_INDEX_DESCRIPTOR));
 
+        // Store this because when the listener we register will be called, secure settings will be closed
+        if (ELASTIC_PASSWORD_HASH.exists(settings)) {
+            elasticPasswordHash.set(ELASTIC_PASSWORD_HASH.get(settings));
+            securityIndex.get().addStateListener(this::possiblySetElasticPassword);
```

Comment:
As discussed already on another channel, I think we need to better define the "lifecycle" of this new password setting. Right now, it is: "the auto-configured elastic password becomes effective soon after the Security index is created, hopefully".

We've discussed already that I was thinking that it has to be effective as soon as the node can handle requests. We can tease it apart, but, let me propose another alternative:

Currently, user authentication does not trigger the creation of the Security index. It doesn't have to. But I propose we make this new auto-configured password be different in this regard. When a client uses the auto-configured password to authenticate elastic, we try to create the index (if it doesn't already exist) and set the elastic user password, before we confirm the authentication as successful. If any operation fails, ie index creation, indexing, or racing with another node that sets a different password, authentication fails. If the index already exists, no magic happens.
The benefit of this approach is that the client is never going to receive an authentication failure for the "promised" password. They can still get an error (ie fail to create the index) but this is not technically breaking the promise, the node is broken, for other APIs too It also somewhat solves the issue with a different value across the nodes: only the password of the first successful authentication is going to be valid henceforth (at least it is behavior that we can easily explain in words)
I also think it's also worth it to make the auto-configured password take precedence over the bootstrap.password and keystore.seed (ie they are ignored if there is an auto configured password hash). This is a tangent conversation, and I can be convinced otherwise. My reasoning is that it is confusing to technically have multiple passwords valid for the same user at a point in time.

Let me know what you think.

Reply:
> We've discussed already that I was thinking that it has to be effective as soon as the node can handle requests. We can tease it apart, but, let me propose another alternative:

Yeap, I have the code to validate the password of the elastic user against the hash in the keystore (until it is set in the index) in another branch as we discussed, I'll update this PR shortly

> Right now, it is: "the auto-configured elastic password becomes effective soon after the Security index is created, hopefully".

It's not "created", it's "recovered" but I'm being pedantic :) We've agreed that this won't be the case ( see above ). It is more like  "the auto-configured elastic password is effective immediately and will be moved to the security index after the Security index is recovered, hopefully"

> let me propose another alternative: 
> ...
> They can still get an error (ie fail to create the index) but this is not technically breaking the promise

wouldn't this prevent the user to authenticate with the "promised" password before the node has read the cluster state or if it fails to do so ? 

> solves the issue with a different value across the nodes

I don't think this issue can manifest in packaged installations. It would need two nodes configured to be members of the same cluster (sharing configuration) starting up for the first time at the same time. But: 

- For any subsequent node to join an existing cluster, the first node needs to be up and running (so the password would be already set ). It might be the case that we failed to set the password during the boot of the first node, one can rightfully argue, but
- For any subsequent node to join an existing cluster, the user needs to run an external tool to consume the enrollment token and reconfigure the node. This external tool can very well remove the autoconfiguration.password_hash from that node's keystore so that it wouldn't even try to promote this to the index.